vault/src/main/java/org/opencadc/vault/files/HeadAction.java

```diff
@@ -150,29 +163,70 @@ DataNode resolveAndSetMetadata(boolean includeContentLength) throws Exception {
         
         log.debug("node path resolved: " + node.getName() + " type: " + node.getClass().getName());
 
-        DataNode dn = (DataNode) node;
-        if (includeContentLength) {
-            syncOutput.setHeader("Content-Length", dn.bytesUsed);
-        }
         syncOutput.setHeader("Content-Disposition", "inline; filename=\"" + node.getName() + "\"");
-        syncOutput.setHeader("Content-Type", node.getPropertyValue(VOS.PROPERTY_URI_TYPE));
-        syncOutput.setHeader("Content-Encoding", node.getPropertyValue(VOS.PROPERTY_URI_CONTENTENCODING));
-        if (node.getPropertyValue(VOS.PROPERTY_URI_DATE) != null) {
-            Date lastMod = NodeWriter.getDateFormat().parse(node.getPropertyValue(VOS.PROPERTY_URI_DATE));
-            syncOutput.setLastModified(lastMod);
-        }
+        
+        ResolvedNode ret = new ResolvedNode();
+        ret.node = (DataNode) node;
+        DataNode dn = ret.node;
+        
+        // determine if data node is up to date or we need to find artifact
+        if (nodePersistence.preventNotFound && (dn.bytesUsed == null || dn.bytesUsed == 0)) {
+            VOSURI targetURI = localServiceURI.getURI(node);
+            Transfer pullTransfer = new Transfer(targetURI.getURI(), Direction.pullFromVoSpace);
+            pullTransfer.version = VOS.VOSPACE_21;
+            pullTransfer.getProtocols().add(new Protocol(VOS.PROTOCOL_HTTPS_GET)); // anon, preauth
 
+            VaultTransferGenerator tg = nodePersistence.getTransferGenerator();
+            ret.protos = tg.getEndpoints(targetURI, pullTransfer, null);
+            ret.artifact = tg.resolvedArtifact;
+        }
+        
+        URI contentChecksum = null;
         String contentMD5 = node.getPropertyValue(VOS.PROPERTY_URI_CONTENTMD5);
         if (contentMD5 != null) {
-            try {
-                URI md5 = new URI("md5:" + contentMD5);
-                syncOutput.setDigest(md5);
-            } catch (URISyntaxException ex) {
-                throw new RuntimeException("BUG: invalid " + VOS.PROPERTY_URI_CONTENTMD5 + " value " + contentMD5);
-            }
+            contentChecksum = URI.create("md5:" + contentMD5);
+        }
+        String contentLength = null;
+        if (dn.bytesUsed != null) {
+            contentLength = dn.bytesUsed.toString();
+        }
+        String contentType = node.getPropertyValue(VOS.PROPERTY_URI_TYPE);
+        String contentEncoding = node.getPropertyValue(VOS.PROPERTY_URI_CONTENTENCODING);
+        String clm = node.getPropertyValue(VOS.PROPERTY_URI_CONTENTDATE); // use if present
+        if (clm == null) {
+            clm = node.getPropertyValue(VOS.PROPERTY_URI_DATE);
         }
+        Date contentLastModified = null;
+        if (clm != null) {
+            contentLastModified = NodeWriter.getDateFormat().parse(clm);
+        }
+        if (ret.artifact != null) {
+            // preventNotFound
```

Comment:
This potentially overrides the values set in the previous lines which might not be very obvious. Maybe re-org so that it's more clear where the specific value originates from (artifact first then node). Just a comment.

Reply:
basically, if preventNotFound logic was triggered above and an artifact was found, it's always right to report that metadata as it's the definitive file metadata

the "else" is potentially getting metadata for an old version of the file, which can't practically be detected

I tried several sequences and this was the most clear to me